stdlib/public/core/Sequence.swift

```diff
@@ -428,28 +427,14 @@ extension Sequence {
       return result
     }
 
-    var hitEnd = false
     var iterator = self.makeIterator()
-    while true {
-      guard let element = iterator.next() else {
-        hitEnd = true
-        break
-      }
-      if try isSeparator(element) {
-        if !appendSubsequence() {
-          continue
-        }
-        if result.count == maxSplits {
-          break
-        }
-      } else {
-        subSequence.append(element)
-      }
-    }
-    if !hitEnd {
-      while let element = iterator.next() {
-        subSequence.append(element)
+    while let element = iterator.next() {
+      if try isSeparator(element) && result.count < maxSplits {
```

Comment:
This would evaluate `isSeparator()` on every element, even if we are not going to append anything.